pkg/controller/gc_controller.go

```diff
@@ -90,70 +100,111 @@ func (c *gcController) Run(ctx context.Context, workers int) error {
+	deletionFailure := false
 
-	// GC backup files and associated snapshots/API objects. Note that deletion from object
-	// storage should happen first because otherwise there's a possibility the backup sync
-	// controller would re-create the API object after deletion.
-	for _, backup := range backups {
-		if !backup.Status.Expiration.Time.Before(now) {
-			glog.Infof("Backup %s/%s has not expired yet, skipping", backup.Namespace, backup.Name)
-			continue
+	for _, volumeBackup := range backup.Status.VolumeBackups {
+		glog.Infof("Removing snapshot %s associated with backup %s", volumeBackup.SnapshotID, kube.NamespaceAndName(backup))
+		if err := c.snapshotService.DeleteSnapshot(volumeBackup.SnapshotID); err != nil {
+			glog.Errorf("error deleting snapshot %v: %v", volumeBackup.SnapshotID, err)
+			deletionFailure = true
 		}
+	}
 
-		// if the backup includes snapshots but we don't currently have a PVProvider, we don't
-		// want to orphan the snapshots so skip garbage-collection entirely.
-		if c.snapshotService == nil && len(backup.Status.VolumeBackups) > 0 {
-			glog.Warningf("Cannot garbage-collect backup %s/%s because backup includes snapshots and server is not configured with PersistentVolumeProvider",
-				backup.Namespace, backup.Name)
-			continue
+	// If applicable, delete backup & metadata file from object storage *before* deleting the API object
+	// because otherwise the backup sync controller could re-sync the backup from object storage.
+	if deleteBackupFile {
+		glog.Infof("Removing backup %s", kube.NamespaceAndName(backup))
+		if err := c.backupService.DeleteBackupFile(c.bucket, backup.Name); err != nil {
+			glog.Errorf("error deleting backup %s: %v", kube.NamespaceAndName(backup), err)
+			deletionFailure = true
 		}
 
-		glog.Infof("Removing backup %s/%s", backup.Namespace, backup.Name)
-		if err := c.backupService.DeleteBackup(c.bucket, backup.Name); err != nil {
-			glog.Errorf("error deleting backup %s/%s: %v", backup.Namespace, backup.Name, err)
+		if deletionFailure {
+			glog.Warningf("Backup %s will not be deleted due to errors deleting related object storage files(s) and/or volume snapshots", kube.NamespaceAndName(backup))
+		} else {
+			if err := c.backupService.DeleteBackupMetadataFile(c.bucket, backup.Name); err != nil {
+				glog.Errorf("error deleting backup metadata file for %s: %v", kube.NamespaceAndName(backup), err)
+				deletionFailure = true
+			}
 		}
+	}
 
-		for _, volumeBackup := range backup.Status.VolumeBackups {
-			glog.Infof("Removing snapshot %s associated with backup %s/%s", volumeBackup.SnapshotID, backup.Namespace, backup.Name)
-			if err := c.snapshotService.DeleteSnapshot(volumeBackup.SnapshotID); err != nil {
-				glog.Errorf("error deleting snapshot %v: %v", volumeBackup.SnapshotID, err)
+	glog.Infof("Getting restore API objects referencing backup %s", kube.NamespaceAndName(backup))
+	if restores, err := c.restoreLister.Restores(backup.Namespace).List(labels.Everything()); err != nil {
```

Comment:
I'm somewhat concerned about iterating through every restore each time we GC a single backup. Other possible options we may want to consider:
- when creating a restore, set a label (ark-backup=foo), then use that as the label selector here (users could remove/edit it, so it wouldn't be foolproof and would require a periodic re-listing as described below)
- maintain an in-memory map of backup->restores and use that here
- periodically re-list all restores and remove any that the backupLister doesn't have

Reply:
as discussed, I'll file an issue for this.